uncompyle6/parsers/reducecheck/ifelsestmt.py

```diff
@@ -240,6 +240,21 @@ def ifelsestmt(self, lhs, n, rule, tree, tokens, first, last):
                     return True
 
             if first_offset > jmp_target:
+                if (tree[2].kind == 'JUMP_FORWARD'
+                  and if_condition[1].kind == 'jmp_false'
+                  and len(tree[3]) == 1
+                  and tree[3][0].kind == 'suite_stmts'
+                  and len(tree[3][0]) == 1
+                  and tree[3][0][0].kind == 'continue'
+                  and if_condition[1][0].attr == tree[3][0][0][0].attr):
+                    # for ...:
+                    #   if ...:
+                    #     ...
+                    #   else:
+                    #     continue
+
+                    return False
+
```

Comment:
This is good! There are a number of small changes though that would improve readability. 

Things like `tree[3]` or worse `tree[3][0][0]` hard to follow. Previously `tree[3]` was assigned to the variable `else_suite` and the variable  `if_condition[1]` was assigned to the variable `jmp`. 

Also we've changed `==` of an ASTnode so that uses `.kind` in for testing. So `tree[2] == "JUMP_FORWARD"` is the same as `tree[2].kind == "JUMP_FORWARD"`.  Note that in new code, I use black conventions which puts strings in double quotes when that is easily possible as it is here. 

Given the above, in general the pattern I have been using is:

```
hoped_for_nonterminal_name = tree[i]
if tree[i] == "hoped_for_nonterminal_name": 
    # other tests and code
```

Missing in the above is the comment general comment mentioned in #411 in words what the intent is. Also missing is the check that the nonterminal we want to reduce on is the kind of "if" that appears inside a loop statement. 

So given all of this here is how I'd write the above: 

```
            if first_offset > jmp_target:
                # A backward or loop jump from the end of an "else"
                # clause before the beginning of the "if" test is okay
                # only if we are trying to match or reduce an "if"
                # statement of the kind that can occur only inside a
                # loop construct.
                if lhs in ("ifelsestmtl", "ifelsestmtc"):
                    jump_false = jmp
                    if (
                        tree[2].kind == "JUMP_FORWARD"
                        and jump_false == "jmp_false"
                        and len(else_suite) == 1
                    ):
                        suite_stmts = else_suite[0]
                        continue_stmt = suite_stmts[0]
                        if (
                            suite_stmts == "suite_stmts"
                            and len(suite_stmts) == 1
                            and continue_stmt == "continue"
                            and jump_false[0].attr == continue_stmt[0].attr
                        ):
                            # for ...:
                            #   if ...:
                            #     ...
                            #   else:
                            #     continue
                            return False
                return True
```

Do you mind revising the above?

Reply:
>                 if lhs in ("ifelsestmtl", "ifelsestmtc"):

This makes the code fail: `lhs` is `ifelsestmt` both times this code section is attempted, as discussed in the issue.

Reply:
>                 if lhs in ("ifelsestmtl", "ifelsestmtc"):

This makes the code fail: `lhs` is `ifelsestmt` both times this code section is attempted, as discussed in the issue.

Reply:
Looks like the "ifelsestmtc" isn't hooked into stmts, in Python 2.7. So this patch hooks it in:
[for-else-continue-2.4.txt](https://github.com/rocky/python-uncompyle6/files/9736476/for-else-continue-2.4.txt)

Reply:
> Looks like the "ifelsestmtc" isn't hooked into stmts, in Python 2.7. So this patch hooks it in: [for-else-continue-2.4.txt](https://github.com/rocky/python-uncompyle6/files/9736476/for-else-continue-2.4.txt)

Your changes change drastically what is happening: Now the change I made in `ifelsestmt()` is rendered useless:
```python
[...]
L. 12:  19-37  ifelsestmtc ::= testexpr c_stmts_opt JUMP_FORWARD else_suite come_froms (14)
L. 12:  19-37  ifelsestmt ::= testexpr c_stmts_opt JUMP_FORWARD else_suite come_froms (14)
Reduce ifelsestmt invalid by check
L. 12:  19     stmt ::= ifelsestmtc (14)
[...]
```

Reply:
> ```python
> Reduce ifelsestmt invalid by check
> ```

I don't see a problem with this. `ifelsestmt` is invalid here. Instead `ifelsestmtc` should match instead. 

If there is a specific problem with decompilation, create a little test case that shows this.

Reply:
> Since your fix resolved the issue, shouldn't yo uauthor the changes rather than me? You can pick my test files up if you like for integration tests.

With respect to who is marked on fixing for this particular issue, I don't really care. My goal here was to get you more educated on this and this was a faster way to show you how things work.

Reply:
> > ```python
> > Reduce ifelsestmt invalid by check
> > ```
> 
> I don't see a problem with this. `ifelsestmt` is invalid here. Instead, `ifelsestmtc` should match.
> 
> If there is a specific problem with decompilation, create a little test case that shows this.

I made the test by reverting my changes, hence nullifying the changes I brought in this PR.
You patch did all the work and both my test cases decompile.

Reply:
I don't mind giving you credit for work I did, but if you feel bad about this, I'll push a new branch with the changes. And you can rebase this PR on top of that. Let me know if that's what you want me to do.

Reply:
My point was: there is no need to modify the function I was trying to change anymore.
The fix is the patch you provided.

I changed authorship for you in the related commit of this PR (I could do that since it appears you do not sign yours with GPG over here).

I changed my additional commit to merely include tests, but I'm yet unsure about their contents.